version-upgrade-operator/README.md

```diff
@@ -0,0 +1,673 @@
+
+  # serviceAccountName
+  #
+  # Service Account name in the `.metadata.namespace` to perform upgrades with.
+  #
+  # The question is, even if we grant customers an admin access, do we still consider
+  # them and us as separate tenants? If so, then we have kind of multi-tenancy, and since
+  # it would be nice to implement it with the Kubernetes RBAC, we need a SA, the same way Flux
+  # does it.
+  #
+  # If we do not care, we can allow the controller to always reconcile resources with its
+  # default permissions, presumably the `cluster-admin`.
+  serviceAccountName: "automation"
+
+  # interval
+  #
+  # Reconciliation interval.
+  interval: 5m
+
+  # targets
+  #
+  # A list of targets to upgrade. They have to be in the same namespace. This can be either specific
+  # targets listed by kind and name or wildcards using label selectors.
+  # Name and labelSelector are mutually exclusive.
+  #
+  # Each target defines a path to its version holding property that should be updated.
+  targets:
+    - kind: App
+      field: spec.version
+      name: cluster3-trivy
+    - kind: HelmRelease
+      field: spec.chart.spec.version
+      name: cluster1-trivy
+    - kind: OCIRepository
+      field: spec.ref.semver
+      name: trivy
+    - kind: HelmRelease
+      field: spec.chart.spec.version
+      labelSelector:
+        app: trivy
+        stage: production
+        vucUpdate: true
+
+  # versionSourceRef
+  #
+  # TODO: should this be cross-namespace?
+  versionSourceRef:
+    kind: ImagePolicy
+    name: trivy-rc
+    namespace: org-testorg # optional; defaults to the same namespace as the VUC CR
+
+  # defaultVersion
+  #
+  # (Optional) When the operator detects a new matching target that has no version set or the version is `latest`, the version
+  # will be set to `defaultVersion`, even outside an upgrade version. This happens only once, when a resource is
+  # discovered for the first time.
+
+  defaultVersion: "1.0.0"
+
+  # versionUpgradeScheduleRef
+  #
+  # Source of the schedule information. This is a reference to a `VersionUpgradeSchedule` CR in the same namespace.
+  # TODO: should this be cross-namespace?
+  versionUpgradeScheduleRef:
+    name: upgrade-stable-at-night
+
+  # versionLock (still being discussed)
+  #
+  # Decides policy towards version changes of targeted objects, that were not introduced by the controller.
+  #
+  # It could be something simple like `True`, which means we lock current version, or `False`,
+  # which means we don't. But it can be more sophisticated, say for example, it is an enum field
+  # with these values:
+  #
+  # "None" - user may change version of the app, controller won't react. When reconciliation time comes,
+  #          the target field version will be set by the controller.
+  # "Corrective" - user may change version of the app, but during reconciliation, the controller will
+  #                restore last version to it, but takes no responsibility for the damage that switching
```

Comment:
So you mean it will correct it only during the upgrade window or even before the upgrade window?

How do we detect the drift? How do we know what was the version applied in the last upgrade window? What about the 1st upgrade window (tho that ties back to the initial CR creation issue I mentioned above)?

> but takes no responsibility for the damage

I get it but dont get it. What kind of responsibility the operator takes lets say even during the upgrade window?

Reply:
The rough idea for corrective behaviour could be describe with these steps:
- controller detects current version of an app and saves it in the VUC status.
- VUC gets reconciled at specified intervals and part of the reconciliation is the corrective behaviour. If now is outside scheduled window and the app version mismatches the one saved in status, the controller corrects it.
- if now is inside scheduled window, there is no point in correction in general, I think, for the usual reconciliation and acting according to the policy is the actual correction then.